### PR TITLE
ldc: split includes to separate output

### PR DESCRIPTION
**Pull request by CyberShadow**
## Description of changes

- Fixes trivial cases of #324306.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by CyberShadow** [comment]
```
19 packages built:
btdu cheesecutter dstep dtools dub gtkd inochi-creator inochi-session ldc ldc.include literate luneta onedrive onedrivegui onedrivegui.dist rund sambamba serve-d tilix
```

**Comment by jtbx** [comment]
I'm a bit torn on this, what's the point if the default output can't compile anything useful? I think the difference in closure size is negligible.

**Comment by CyberShadow** [comment]
> I'm a bit torn on this, what's the point if the default output can't compile anything useful?

What do you mean? The default output references the `include` output.

> I think the difference in closure size is negligible.

It's not:

```console
$ nix path-info -S "$(nix-build --no-out-link -E '(import <nixpkgs> {}).callPackage ./package.nix {}')"
/nix/store/c7d1kz9rlj38k9w8n8smmp6msq2x0994-ldc-test	 1257047336

$ nix path-info -S "$(nix-build --no-out-link -E '(import ../../../work/extern/nixpkgs {}).callPackage ./package.nix {}')"
/nix/store/rsr7y65w6kirsr88idbqn01yhd90dhwp-ldc-test	   78766032
```

**Comment by CyberShadow** [comment]
To clarify what's going on:

- Programs built by LDC have references to LDC, even without `-g` or `-shared`.
- Without `-g` or `-shared`, these references are all to source files in Phobos. I think they're coming from `__FILE__` instances in exception constructor default arguments.
- LDC itself (the compiler binaries) is big, but it is also pulling in LLVM, which is even bigger.
- By moving the Phobos sources to a separate output (`include`), we can make it so that LDC-built programs only depend on those sources. Those sources in turn don't depend on anything, so we can avoid pulling in LLVM etc.
- Nothing changes from LDC's perspective, other than that the Phobos sources are at a new location. The output containing the compiler binaries now also references the output containing the includes.
- There is room for more improvements to do the same thing for `-shared` builds, if we can move the shared libraries to their own separate output as well. However, that one doesn't seem to be quite as simple as for `include`.

**Comment by jtbx** [comment]
Ah that makes sense now, thanks for all that.

> To clarify what's going on:
> 
>  * Programs built by LDC have references to LDC, even without `-g` or `-shared`.
> 
>  * Without `-g` or `-shared`, these references are all to source files in Phobos. I think they're coming from `__FILE__` instances in exception constructor default arguments.
> 

Yep I think so too, this was discussed in #288841

**Review by jtbx — APPROVED** [review]

**Comment by CyberShadow** [comment]
Added two tweaks:
1. Moved `.d` files into `include/d/` so as to not make a mess with `buildEnv` et al
2. Added a test

**Comment by lionello** [comment]
See https://github.com/NixOS/nixpkgs/pull/111525 which had fixes for `dmd`.

**Comment by CyberShadow** [comment]
Thanks, `removeReferencesTo` is a rather blunt tool which I'd rather leave as a last resort.

**Comment by lionello** [comment]
@CyberShadow I'm not sure you can avoid `removeReferencesTo` because IIRC each exception has the filename in it, which means Nix will keep references to bootstrap and/or Phobos around.

**Comment by CyberShadow** [comment]
Well, I haven't looked at the situation with `dmd` / `dmd-bootstrap` specifically. What I did here was to put the Phobos sources in a separate output, which means that programs still refer to those sources, but that is as far as the dependency chain goes, so no more references to LLVM / bootstrap LDC / etc from built programs. Maybe something like that is possible for dmd-bootstrap too, if we tweak its `dmd.conf`?

File: pkgs/by-name/ld/ldc/package.nix [hunk 1]
```diff
@@ -142,4 +148,34 @@ stdenv.mkDerivation (finalAttrs: {
     maintainers = with maintainers; [ lionello jtbx ];
     platforms = [ "x86_64-linux" "i686-linux" "aarch64-linux" "x86_64-darwin" "aarch64-darwin" ];
   };
+
+  passthru.tests = let
```
Comment (pbsds) [inline pkgs/by-name/ld/ldc/package.nix:152]:
```suggestion
  passthru.tests = let
    ldc = finalAttrs.finalPackage;
```

File: pkgs/by-name/ld/ldc/package.nix [hunk 2]
```diff
@@ -1,6 +1,7 @@
 { lib
 , stdenv
 , fetchFromGitHub
+, ldc
```
Comment (pbsds) [inline pkgs/by-name/ld/ldc/package.nix:4]:
```suggestion
```

File: pkgs/by-name/ld/ldc/package.nix (hunk 1 above)
Comment (CyberShadow) [reply pkgs/by-name/ld/ldc/package.nix:151]: Thanks!

**Pull request merged**